src/libraries/System.Private.CoreLib/src/System/Double.cs

```diff
@@ -440,6 +440,162 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
+                ret = (ulong)(long)(val - two63) + 0x8000000000000000UL;
+            }
+            return ret;
+        }
+
+        [StackTraceHidden]
+        private static int ConvertToInt32Overflow(double val)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (val is > -IntMaxValueOffset - 1 and < IntMaxValueOffset)
+            {
+                int ret = (int)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static uint ConvertToUInt32Overflow(double val)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (val is > -1.0 and < UIntMaxValueOffset)
+            {
+                uint ret = (uint)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static long ConvertToInt64Overflow(double val)
+        {
+            const double two63 = IntMaxValueOffset * UIntMaxValueOffset;
+
+            // Note that this expression also works properly for val = NaN case
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (val is > -two63 - 0x402 and < two63)
+            {
+                long ret = (long)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static ulong ConvertToUInt64Overflow(double val)
+        {
+            const double two64 = UIntMaxValueOffset * UIntMaxValueOffset;
+            // Note that this expression also works properly for val = NaN case
+            if (val is > -1.0 and < two64)
+            {
+                ulong ret = (ulong)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        private static ref double DoubleConversionTable => ref MemoryMarshal.GetReference([0.0, UIntMaxValueOffset]);
```

Comment:
Nit: Does this need to be factored into a helper method? Can it be inline?

Reply:
Ah ok, you are using it in two places.

Reply:
Two methods use this so I wanted to avoid duplicating it in both.